coffeescript-loader/loader.js

```diff
@@ -25,41 +34,50 @@ export function resolve(specifier, context, defaultResolve) {
   return defaultResolve(specifier, context, defaultResolve);
 }
 
-export function getFormat(url, context, defaultGetFormat) {
+export async function load(url, context, defaultLoad) {
   // Now that we patched resolve to let CoffeeScript URLs through, we need to
   // tell Node.js what format such URLs should be interpreted as. Because
   // CoffeeScript transpiles into JavaScript, it should be one of the two
   // JavaScript formats: 'commonjs' or 'module'.
   if (extensionsRegex.test(url)) {
-    // CoffeeScript files can be either CommonJS or ES modules, so we want any
-    // CoffeeScript file to be treated by Node.js the same as a .js file would
-    // be at the same location (based on the "type" field in the nearest
-    // parent package.json file). To determine how Node.js would interpret an
-    // arbitrary .js file, append .js to our CoffeeScript URL and see what
-    // format Node.js returns for such a URL. If 'commonjs' is returned, a
-    // handler will need to be registered with require.extensions to process
-    // that file via the CommonJS loader.
-    const { format } = defaultGetFormat(`${url}.js`);
-    return {format};
-  }
+    // CoffeeScript files can be either CommonJS or ES modules, but since load
+    // handles both format and source based on the same url, it cannot be used
+    // for non-js files. Instead, a quick search up the filesystem for a
+    // package.json with a `type` field settles the format issue.
+    const format = await getPackageType(dirname(fileURLToPath(url)));
 
-  // Let Node.js handle all other URLs.
-  return defaultGetFormat(url, context, defaultGetFormat);
-}
+    // source is ignored (never checked) for cjs, so safe to omit
+    if (format === 'commonjs') return { format };
 
-export function transformSource(source, context, defaultTransformSource) {
-  const { url, format } = context;
+    const { source: rawSource } = await defaultLoad(url, { format });
+    // This hook converts CoffeeScript source code into JavaScript source code
+    // for all imported CoffeeScript files.
+    const transformedSource = CoffeeScript.compile(''+rawSource, {
+      bare: true,
+      filename: url,
+    });
 
-  // This hook converts CoffeeScript source code into JavaScript source code
-  // for all imported CoffeeScript files.
-  if (extensionsRegex.test(url)) {
     return {
-      source: CoffeeScript.compile(source, {bare: true, filename: url})
+      format,
+      source: transformedSource,
```

Comment:
`rawSource` (as returned by `defaultLoad()`) is a string

Reply:
>I’m not sure we need to make this section more verbose.

Sure. which part would you like removed—the `transformedSource` constant? That seemed to be worthwhile to explicitly call out (plus is a bit cleaner and easier to debug), but I don't have particularly strong feelings about it. But please do let me know if that is the bit you'd like consolidated.

>Is rawSource not already a string? Why do we need to cast it?

That's what `''+` does 🙂 last I checked, it was the most efficient way (by a lot) to cast to a string.

Reply:
If you think `transformedSource` makes it clearer, then sure, it’s worth the added lines.

Yes either `''+rawSource` or `` `${rawSource}` `` will cast `val` into a string, but why is either approach necessary? Is `rawSource` not already a string?

Reply:
Because it can also be multiple kinds of Array: [docs/ESM → getSource](https://nodejs.org/api/esm.html#esm_getsource_url_context_defaultgetsource) (the return types haven't changed)

Reply:
I’m wondering when it _wouldn’t_ be a string; maybe only for wasm? But I guess for an example it’s better to do `${rawSource}` or `rawSource.toString()` to be really explicit and hint that the loader author shouldn’t assume that `rawSource` is necessarily always a string.

Reply:
👍 explicit is better 🙂

As for when, I'm not sure of the conditions off the top of my head. I do know that when I was debugging the import CJS issue, some `source`s were buffers and not strings (and some were strings). I remember the CoffeeScript lib import was a string, and I'm ~80% sure this loader's test fixture was a buffer. If you'd like to know for sure, I can add back the logging and check in the morning.

Reply:
> If you’d like to know for sure, I can add back the logging and check in the morning.

No, please don’t spend any more time on this, it’s not worth the effort. I remember Bradley saying something about the source of a wasm import always being a buffer, so I bet it has to do with that (at least).